### ISSUE TITLE
Inner and outer product should accept tensor products of bra-kets

### ISSUE THREAD

**Issue by cshaa** [sympy/sympy#19540]
## Input code
```python
k1 = OrthogonalKet(1)
k2 = OrthogonalKet(2)

lh = TensorProduct(k1, k2)           # |1>|2>
rh = TensorProduct(k1 + k2, k1 - k2) # |1>|1> - |1>|2> + |2>|1> - |2>|2>

prod = InnerProduct(Dagger(lh), rh)
srepr(qapply(prod))
```
## Expected result
```python
# - <2|<1|1>|2>
Mul(Integer(-1), InnerProduct(
  TensorProduct(OrthogonalBra(1), OrthogonalBra(2)),
  TensorProduct(OrthogonalKet(1), OrthogonalKet(2))
))
```
## Actual output
```
TypeError: KetBase subclass expected, got: (|1> + |2>)x(|1> - |2>)
```
## Rationale
When working on a product space H⊗H, a tensor product of kets represents a state, much like one ket represents a state in H. Therefore, all the things we can do on bra-kets, we should be able to do on their tensor products.

**Comment by IvyMercyLyons** [comment]
Is it fine if I address this?

**Comment by oscarbenjamin** [comment]
Yes,  go ahead. I don't know the quantum code so well myself but maybe @m93a could review a pull request.

**Comment by IvyMercyLyons** [comment]
I will mention first that this is my first involvement in a GitHub project, so I may misunderstand some aspects of the social protocols for comment-section usage, although they  mostly seem to be clear so far!

I have searched through the project for items related to the tensor product and found that a number of packages contain their own definitions for the tensor product. I have also reviewed the history of these coexisting definitions and found that people have attempted to merge or rename them, though in some cases perhaps the apparent magnitude of the task discouraged people and they did not complete it. Although this is not quite the same context as the finite-dimensional one in these other packages, I am thinking I could perhaps attempt to utilize the existing frameworks for tensor products, tensors and traces/contractions to create a consistent handling that is simply adapted to the infinite-dimensional case - part of this is to account for the fact that tensors may take multiple "arguments," and there is no way to specify "how" the tensors in the inner product may be interacting. Is it okay if I do this? If the words are not meaningful, the main idea is that I want to modify the way that a number of major functions work in "quantum." (I apologize if I'm not supposed to ask about these things in advance like this, I am still learning.)

I had two more questions about more extreme details:

- The definition of the above "InnerProduct" inherits from the basic expression class, rather than the quantum expression class, and I notice that this fact is often important in the code, as well as the fact that all quantum expression descendants are assumed not to commute (the InnerProduct is assumed to only produce scalars, which is precisely the kind of behavior I would be generalizing.) Is it okay if I modify the InnerProduct to inherit from the quantum expression class, account for this in the code, and create handling for noncommutative quantum expression descendants?

- Would it make sense for me to later possibly contribute a functional analysis module (considering the goal, probably not so large upon its creation,) so that the extra definitions in the quantum module may be less extensively specialized, or is the quantum library meant to be somewhat autonomous?

**Comment by m93a** [comment]
I'll be speaking for myself – and while I am a contributor, my contributions were simply a result of bodging things together until, so I don't have a great deal of understanding sympy as a whole. So my opinions might not be the best advice, but I'll share them anyway 😁️

1) Since the inner product of `<a|` and `|b>|c>` is a perfectly valid expression that results in `(<a|b>) |c>`, changing `InnerProduct` to inherit from QuantumExpr sounds like a reasonable thing to do. AFAIK it should still be possible to set `is_commutative` manually for the occasions when the result really is a scalar.

2) A functional analysis module sounds like a great idea to me, since it's really hard to do anything smart with the Dirac notation unless you know the properties of the operators. I think that that the FA module should be as autonomous from the core sympy, as physics.quantum is now. But making physics.quantum dependent on the FA module would be very beneficial for it IMO.

**Comment by IvyMercyLyons** [comment]
How many changes would be suitable as an address to the specific issue? A lot of the improvements would, I suppose, be out of the scope of the original problem, which can be solved pretty simply. Should those be included into an independent pull or should this also have changes that, for example, would contribute to the general development of the InnerProduct function (eg, general inner product materials in a separate package, arguments where the significant "indices" would be described, etc?)

On the InnerProduct itself - I notice it calls an _eval_innerproduct function in the bras and kets, which themselves refer to a "dispatch function" in the qexpr module. This would not generalize really well to arbitrary objects or inner products themselves. Would it be considered too extreme to have the InnerProduct function be entirely self-contained? I checked other SymPy packages to see how common this kind of arrangement is and it is not unique to this, but here it could hold the development of the general inner product back a bit (and though I can see in this form that it could have value to the creation of special-case kets and bras, I am thinking of including a simple way to define to some extent the contents of those kets and bras, which is currently not possible (and which is one of the reasons I took a while to respond this time :stuck_out_tongue: .)) To reiterate what I mean - rather than calling the _eval_innerproduct in the "states" module, the InnerProduct function would be as self-contained as possible and only make external references to obtain information for the product itself, not for computations themselves.

**Comment by oscarbenjamin** [comment]
I'm not sure I understand your question but I will say this: Currently the quantum module is effectively unmaintained. This means that there might not be anyone who can advise on what is a good change to make or not for the broader direction of the module. I personally can't advise although what you say sounds reasonable to me. Hopefully @m93a will be able to help...

**Comment by m93a** [comment]
> How many changes would be suitable as an address to the specific issue?

If it's up to me, there's no upper bound 😁️ This module needs some love and care, and if you want to rewrite some parts to make it more powerful/manageable, go for it. I'll gladly read through the code and provide a second opinion. Just be sure to chunk up your work to reasonably-sized PRs.

> Would it be considered too extreme to have the InnerProduct function be entirely self-contained?

As far as I know, right now the only internal function that redefines `_eval_innerproduct` is `OrthogonalKet`, so moving that logic elsewhere shouldn't be a problem. Currently, the quantum module is half-baked in a way that expects you to define new classes that extend Ket for different purposes, which isn't a good API I'd say.

I'm not sure if there are any “ket classes”, other than OrthogonalKet, that might be useful. I think that vast majority of problems in QM are described with orthonormal basis made from the eigenstates of some operator. So limiting the ket types to “general” and “orthogonal” sounds reasonable… But if you find a good way to make it extendable even after moving the logic, that's great!

The way I think about it, the best API for bra-kets would let you define an operator and then construct an orthogonal basis from it. Then you'd have a “referernce frame” and you could do basis transformations (automatic Clebsch-Gordan?), matrix representations of other operators (albeit possibly infinite-dimensional), ladder operators and things like that. However making this work sounds like huge undertaking, *if* it's possible at all.

**Comment by ellisonbg** [comment]
It shouldn't be too difficult to enable inner and outer products to work with tensor products of states. For inner products it is mostly about carrying around the tensor products and then reducing them to their individual inner products. For outer products it is a bit more complicated as qapply will also need to be modified to handle the outer products of tensor products.
